Prometheus & Grafana

elipe17 opened the pull request:
## Summary of Changes
- Added local MVP for Prometheus monitoring/metric collection
- Added local MVP for Grafana dashboarding
Pull request closes #3044 

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up --build
```

2. Open http://localhost:3000/ and sign in, make sure you're sys admin or a developer
3. Submit some files to generate postgres, django, and celery metrics
4. Navigate to http://localhost:3000/grafana/. User/Pass: admin/admin(note for some reason the proxy does not work the same as Kibana. A follow on issue IMO)
5. You can also open localhost:9090 if you want to try messing around with Prometheus.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [ ] Prometheus is running in local dev environment
- [ ] Prometheus UI shows all the apps and services running
- [ ] Grafana dashboard is accessible in local dev environments
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Added minimum config for prometheus

elipe17 committed:
- Added exporters for celery and postgres
- Added Grafana and dashboards

elipe17 committed:
- moved plg config to appropriate directories
- Updated frontend to allow navigation to Grafana

elipe17 committed:
- add todo

elipe17 committed:
Merge branch 'develop' into 3044-prometheus-grafana

elipe17 committed:
- Update dashboard

jtimpe commented on tdrs-backend/tdpservice/settings/common.py:
```diff
@@ -105,7 +110,7 @@ class Common(Configuration):
     DATABASES = {
         "default": {
-            "ENGINE": "django.db.backends.postgresql",
+            "ENGINE": "django_prometheus.db.backends.postgresql",
```
another thing we can consider is redis monitoring - https://github.com/korfuri/django-prometheus/tree/master?tab=readme-ov-file#monitoring-your-caches

also cool that we can add our own metrics - https://github.com/korfuri/django-prometheus/tree/master?tab=readme-ov-file#adding-your-own-metrics

andrew-jameson replied:
Do we need to enable a prometheus-specific logger? I'm poking around the dashboard and only see metrics, no logs.

edit: Is it worth pulling in #3047 into scope for this PR? I didn't realize those ACs weren't part of this.

elipe17 replied:
Prometheus is not a log aggregator it is for metric collection and time series analysis. We need to integrate Promtail and Loki if we want to capture tdp logs and other dependency logs.

raftmsohani replied:
we can use telegraf, but Promtail is the official log tailer for Grafana.

jtimpe commented on tdrs-backend/tdpservice/settings/common.py:
```diff
@@ -531,3 +536,4 @@ class Common(Configuration):
     IGNORE_DUPLICATE_ERROR_PRECEDENCE = os.getenv("IGNORE_DUPLICATE_ERROR_PRECEDENCE", False)
     BULK_CREATE_BATCH_SIZE = os.getenv("BULK_CREATE_BATCH_SIZE", 10000)
     MEDIAN_LINE_PARSE_TIME = os.getenv("MEDIAN_LINE_PARSE_TIME", 0.0005574226379394531)
+    BYPASS_OFA_AUTH = os.getenv("BYPASS_OFA_AUTH", False)
```
i was going to comment on the variable naming but i see that this also replaces `BYPASS_KIBANA_AUTH`, which would have been my suggestion. good work 👍

jtimpe commented on tdrs-backend/Pipfile:
```diff
@@ -59,6 +59,7 @@ cerberus = "==1.3.4"
 xlsxwriter = "==3.1.9"
 openpyxl = "==3.1.2"
 sendgrid = "==6.10.0"
+django_prometheus = "==2.3.1"
```
any concerns with this third party lib? i took a quick look through the repo and everything seems pretty good, it just doesn't appear to be officially maintained.

elipe17 replied:
Good question, everywhere I looked it was the dependency that everyone recommended/talked about. As far as I know there isn't a django plugin maintained by the prometheus community.

raftmsohani replied:
Why not using python library instead? you will only have to add the line in middleware but it is well maintained

jtimpe reviewed (approved):
Works great! Seeing the metrics come in locally is pretty cool and i anticipate we'll get quite a lot of use out of it. A couple of small comments but overall approved 🚀

andrew-jameson commented on tdrs-backend/tdpservice/users/api/authorization_check.py:
```diff
@@ -65,9 +65,29 @@ def get(self, request, *args, **kwargs):
+
+        user_in_valid_group = user.is_ofa_sys_admin
+
+        if (user.hhs_id is not None and user_in_valid_group) or settings.BYPASS_OFA_AUTH:
```
This feels problematic as it will lock our dev team out of the monitoring tool for prod. Do we want to simply utilize the default administrator user for our use?

Also, perhaps out of scope for this ticket but I know Sentry is going to roll up information from lower environments, will we have a grafana per instance? If not, development team will need a way to view information on staging and lower environments. Perhaps we need to architect this next bit out in the cloud ticket #3046 .

elipe17 replied:
The default admin login is still required for this. Perhaps we should only ensure the user is in the correct group and not worry about an HHS ID. That would resolve prod loss of access while also keeping our Grafana instance gated behind our nginx proxy.

I don't see why we couldn't have a single Grafana for all envs. We just need to specify the URL for nginx to proxy pass to it.

raftmsohani replied:
I guess we can also only have one prometheus instance?

raftmsohani commented on tdrs-backend/docker-compose.yml:
```diff
@@ -69,6 +69,52 @@ services:
+
+  celery-exporter:
+    restart: always
+    image: ovalmoney/celery-exporter:1.5.1
```
This is an interesting exporter, really like the way it exports the metrics. How are we planning to implement this in cloud.gov? dockerized? I guess it should sit with redis?

raftmsohani commented on tdrs-backend/docker-compose.yml:
```diff
@@ -69,6 +69,52 @@ services:
+    depends_on:
+      - web
+
+  postgres-exporter:
```
I am curious to see if this works in cloud.gov. I played a bit with ssh into cloud.gov and tried to connect it to the DB but failed. will spend more time later, but do you think we can connect to cloud.gov DB?

elipe17 replied:
LEts keep PLG its own separate app/deployment. All compose stuff goes in the `plg` dir and it gets its own manifest....

raftmsohani commented on tdrs-backend/tdpservice/users/api/authorization_check.py:
```diff
@@ -65,9 +65,29 @@ def get(self, request, *args, **kwargs):
+        """Handle get request and verify user is authorized to access kibana."""
+        user = request.user
+
+        user_in_valid_group = user.is_ofa_sys_admin
```
Use user group cred control. we can add can_use_grafana and give/revoke access for any user from admin page

elipe17 committed:
- Updated Grafana view to allow access based on group only since devs need it too

elipe17 committed:
- remove setting

elipe17 commented on tdrs-backend/tdpservice/users/api/authorization_check.py:
```diff
@@ -65,9 +65,29 @@ def get(self, request, *args, **kwargs):
+        """Handle get request and verify user is authorized to access grafana."""
+        user = request.user
+
+        user_in_valid_group = user.is_ofa_sys_admin or user.is_developer
```
Since we plan to only have a single grafana instance, we can access it in the envs where the Developer group is defined/used. We dont need to be in prod to see it

elipe17 committed:
- Update frontend vars

elipe17 committed:
- Add deploy script for pg-exporter

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 3044-prometheus-grafana

elipe17 committed:
- Update lock file

elipe17 committed:
- Added promtail and loki
- Added loki dashboard and update plg resources
- Added backend file logger
- UPdated nginx to write logs to file

elipe17 committed:
- Gave datasources UIDs
- Updated logs dashboard with correct loki UID

raftmsohani reviewed (approved):
LGTM!

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 3044-prometheus-grafana

elipe17 committed:
- Update lockfile

andrew-jameson reviewed (approved):
Approved via #3174

elipe17 committed:
Merge pull request #3174 from raft-tech/3047-plg-mvp

Local PLG MVP

elipe17 committed:
Merge branch 'develop' into 3044-prometheus-grafana

elipe17 committed:
Merge branch 'develop' into 3044-prometheus-grafana

jtimpe commented on tdrs-backend/tdpservice/settings/common.py:
```diff
@@ -228,30 +233,35 @@ class Common(Configuration):
         "loggers": {
             "tdpservice": {
-               "handlers": ["application"],
+               "handlers": ["application", "file"],
```
i believe this to have broken deployments - https://app.circleci.com/pipelines/github/raft-tech/TANF-app/28966/workflows/bbe12d8d-d147-4288-8308-e9e8a6ec6e22/jobs/85249

elipe17 replied:
Indeed it has. I changed the path for the log file on my PLG deployment ticket because of this issue. I Will make a hotfix PR. Thank you for catching this!
